XenAdmin/Alerts/Types/ClientUpdateAlert.cs

```diff
@@ -134,17 +135,64 @@ public static void DownloadAndInstallNewClient(ClientUpdateAlert updateAlert, IW
             {
                 Process.Start(outputPathAndFileName);
                 log.DebugFormat("Update {0} found and install started", updateAlert.Name);
-                downloadAndInstallClientAction.ReleaseInstaller();
+                downloadAndInstallClientAction.ReleaseDownloadedContent();
                 Application.Exit();
             }
             catch (Exception e)
             {
                 log.Error("Exception occurred when starting the installation process.", e);
-                downloadAndInstallClientAction.ReleaseInstaller(true);
+                downloadAndInstallClientAction.ReleaseDownloadedContent(true);
 
                 using (var dlg = new ErrorDialog(Messages.UPDATE_CLIENT_FAILED_INSTALLER_LAUNCH))
                     dlg.ShowDialog(parent);
             }
         }
+
+        public static void DownloadSource(IWin32Window parent)
+        {
+            string outputPathAndFileName;
+            using (var saveSourceDialog = new SaveFileDialog())
+            {
+                saveSourceDialog.FileName = BrandManager.BrandConsoleNoSpace + "-source.zip";
+                saveSourceDialog.DefaultExt = "zip";
+                saveSourceDialog.Filter = "(*.zip)|*.zip|All files (*.*)|*.*";
+                saveSourceDialog.InitialDirectory = Win32.GetKnownFolderPath(Win32.KnownFolders.Downloads);
+
+                if (saveSourceDialog.ShowDialog() != DialogResult.OK)
+                {
+                    return;
+                }
+                outputPathAndFileName = saveSourceDialog.FileName;
+            }
+
+            var clientVersion = Updates.ClientVersions.FirstOrDefault();
+            if (clientVersion == null) { return; }
+            var sourceUrl = clientVersion.SourceUrl;
+            var downloadSourceAction = new DownloadSourceAction(clientVersion.Name, clientVersion.Version, new Uri(sourceUrl), outputPathAndFileName);
+
+            using (var dlg = new ActionProgressDialog(downloadSourceAction, ProgressBarStyle.Continuous))
+            {
+                dlg.ShowCancel = true;
+                dlg.ShowDialog(parent);
+            }
+
+            if (!downloadSourceAction.Succeeded)

```

Comment:
I don't think the rest of this method is needed; we only save the zip, we don't need to close the application or disconnect, hence no need to check for unfinished tasks.